### ISSUE TITLE
Move to QtWebEngine

### ISSUE THREAD

**Issue by piranna** [Coretool/noGUI#13]
I think we need to move to [QtWebEngine](http://doc.qt.io/qt-5/qtwebengine-overview.html)... I tried to build `noGUI` the other day and it failed, and I'm not fully sure if it's due to usage of latest version of Qt where `QtWebkit` has been deprecated. On the other hand, `QtWebEngine` has a full instance of Chromium that would give us a better compatibility and to work in pair with the latest PhantomJS, and also seems to be easy to add an [interceptor](http://doc.qt.io/qt-5/qwebengineurlrequestinterceptor.html#details) that would allow us to remove the header that blocks the access to Google, Facebook, GitHub, StackOverflow and others from iframes if we find this is needed.

**Comment by Coretool** [comment]
Yeah, we should move. Coding with QtWebEngine doesn't seem to be much different from QtWebkit. 

> [...] and work in pair with the latest Phantomjs

Does this mean we can use their build scripts (again) ^^ ?

I think the inveptor makes sense since a NodeOS (with GUI) can surely be a every day os !

**Comment by piranna** [comment]
> Yeah, we should move. Coding with QtWebEngine doesn't seem to be much
> different from QtWebkit.

I don't think so, main problem would be to compile it statically, but seems
PhantomJS fixed it.

> > [...] and work in pair with the latest Phantomjs
> 
> Does this mean we can use their build scripts (again) ^^ ?

No... but we can use them as reference about how to do the things, the same
way I used Cross Linux from Scratch to create the build process of NodeOS
:-)

> I think the inveptor makes sense since a NodeOS (with GUI) can surely be
> a every day os !

Yes, but mostly because it's the most simple way to enable that
functionality, and also there's plugins for the browsers for that, so we
can still use them for the GUI. Has anyone tried to use os.js on NodeOS?

**Comment by Coretool** [comment]
So the guys from phantomjs have got a way to build it statically (?) AFAIK they only use QtWebkit... There's even a prototype for CEF but no one for QtWebEngine

Haven't tried it on neither NodeOS nor noGui...

**Comment by piranna** [comment]
> So the guys from phantomjs have got a way to build it statically (?)
> That's good news!

PhantomJS provides a sonewhat statically linked binary. With PhantomJS 1
there was no problem because they used QtWebkit but was too limited, on
PhantomJS 2 they started to use QtWebEngine but there was problems with ICU
and the versions of the C library on Linux, but they found a way to solve
this on PhantomJS 2.1, that's why I said of using their scripts and options
as reference. Could you be able to take a look?

> Haven't tried it on neither NodeOS nor noGui...

We should take a look of it on noGUI first...

**Comment by Coretool** [comment]
They still have QtWebkit in their Src dir (ariya/phantomjs@2.1.1) and their python build script points to QtWebkit too... Is ariya/phantomjs just the wrong repository ?

**Comment by piranna** [comment]
> They still have QtWebkit in their Src dir (ariya/phantomjs@2.1.1) and their python build script points to QtWebkit too... Is ariya/phantomjs just the wrong repository ?

So seems they didn't moved to QtWebEngine yet... :-/ We should consider to move ourselves then, but first let's release a version with QtWebkit first.

**Comment by Coretool** [comment]
Yeah, that's a bit sad :/... I'll look to continue noGUI ... Aurora is also faltering :/

**Comment by Coretool** [comment]
Just as an information, I'll try to build a static QtWebEngine tonight, so fingers crossed ^^

**Comment by piranna** [comment]
Good luck! ;-)

**Comment by piranna** [comment]
Good luck! ;-)

**Comment by Coretool** [comment]
I followed [this SO question](http://stackoverflow.com/questions/30316880/how-to-build-qtwebengine-from-qt-5-5-git-source) As far as I can tell it worked. I'll try it again with qt5.7 in the next couple of days.

**Comment by piranna** [comment]
That's cool! Can you show an output of the ldd command (and a screenshoot
:-P)? Also, instead of using git, could you be able to use tarballs as I
sugested on other issue?

**Comment by Coretool** [comment]
Well I still have to get rid of xcb and build it for the `linuxfb` platform. I'll make a screenshot when I've got that up and running.

Yeah, I'll do that.

**Comment by piranna** [comment]
Get rid of xcb seems trivial, anyway left it by now just for testing. Add
support for framebuffer (it should be similar as how we have it for
QtWebkit) and just give it a test to know it's working. Are you doing your
changes in an independent branch?

**Comment by Coretool** [comment]
I'll push them to an independent branch once I've written the scripts and adjusted the current noGUI code to the `qtwebengine`  (which should be pretty straight forwaed). Adding support for the `linuxfb` works as in the previous version.

**Comment by piranna** [comment]
Fine! :-D Give me the ldd output please, I want to check it.

**Comment by Coretool** [comment]
As soon as I 'm back at my computer ^^

**Comment by piranna** [comment]
ACK

**Comment by Coretool** [comment]
Really bad news. The I built a shared version instead of the static version =/. However, it seems like it's possible to build it static since the build seems to fail at some "minor" errors only (or that's what it fail at now) 

People seem to have had the same problem in the [past](https://forum.qt.io/topic/57888/solved-qt5-5-static-build-failed-qqmldebugserverthread-run-error-undefined-reference-to-qtcpserverconnection-qtcpserverconnection/3). However, the seem to have been able to fix it, so I'll try it again with the instructions on this thread.

**Comment by Coretool** [comment]
Ok. I can retake the claim that I built a static version, finally got it (I guess)

**Comment by Technohacker** [comment]
Move to Qt WebEngine has been done :)

**Comment by Coretool** [comment]
Thanks ^^

**Comment by Technohacker** [comment]
@Coretool I suppose this issue can now be closed :D

**Comment by piranna** [comment]
I've done some fixes on the build script and I'm compiling it just now to check if it works as expected so we can close this. On the other hand, we should check about the interceptor I told on the first comment before clossing this issue, or open a new one specific for it.

**Comment by piranna** [comment]
@Technohacker, has you check your code compiles on a clean slate environment? What's your current system? I have been trying all the day to compile your changes to use QtWebEngine and it doesn't work statically, and with dynamic libraries it's asking me a lot more dependencies...

**Comment by piranna** [comment]
Ok, the `libegl1-mesa-dev` package is needed by Chromium itself, or at least I was not able to configure the `QtWebEngine` module. Knowing that Chromium uses EGL directly, probably would be a better thing to directly use it...

**Comment by piranna** [comment]
```
Project ERROR: Unknown module(s) in QT: qml quick webchannel
```

Definitely, this is not the way to go. Ideally QtWebEngine should be a good option, but I can't be able to replicate it. I've done a pull-request at https://github.com/Coretool/noGUI/pull/22 to see if you can check the fixes and try to find what's happening. Seems probably start from scratch with `content_browser` would be a better idea.

**Comment by Technohacker** [comment]
@piranna Sorry, I couldn't test with a static build of Qt because from my experience with Android, Chromium is a huge project to compile

**Comment by piranna** [comment]
Then you should have adviced it...

**Comment by Technohacker** [comment]
Sorry, my mistake :[
